ethereum/retesteth/src/main/java/tech/pegasys/pantheon/ethereum/retesteth/methods/DebugStorageRangeAt.java

```diff
@@ -0,0 +1,147 @@
+    final int limit = parameters.required(request.getParams(), 4, Integer.class);
+
+    final UInt256 blockId = UInt256.fromHexString(blockHashOrNumber);
+    //    final Optional<TransactionWithMetadata> optional;
+    final Optional<Hash> blockHash;
+    if (blockId.fitsLong()) {
+      blockHash = context.getBlockchainQueries().getBlockHashByNumber(blockId.toLong());
+    } else {
+      blockHash = Optional.of(Hash.fromHexStringLenient(blockHashOrNumber));
+    }
+
+    // TODO deal with in-block locations
+    // if (blockHash.isPresent()) {
+    //  if (context.getBlockchainQueries().getTransactionCount(blockHash.get()) >= transactionIndex)
+    // {
+    //    // post block state
+    //  } else {
+    //    // sub-block state
+    //  }
+    // }
+
+    return extractStorageAt(
+        request,
+        accountAddress,
+        startKey,
+        limit,
+        context
+            .getProtocolContext()
+            .getWorldStateArchive()
+            .getMutable(
+                context
+                    .getBlockchainQueries()
+                    .blockByHash(blockHash.get())
+                    .get()
+                    .getHeader()
+                    .getStateRoot())
+            .get());
+    //    blockHash.map(blockHash -> context.getBlockReplay());
+    //      Optional<BlockWithMetadata<Hash, Hash>> block =
+    // context.getBlockchainQueries().blockByNumberWithTxHashes(blockId.toLong());
+    //      if (block.isPresent()) {
+    //        blockHash = block.get().getHeader().getHash();
+    //        optional =
+    //            context
+    //                .getBlockchainQueries()
+    //                .transactionByBlockNumberAndIndex(blockId.toLong(), transactionIndex);
+    //      } else {
+    //        optional = Optional.empty();
+    //      }
+    //    } else {
+    //      optional =
+    //          context
+    //              .getBlockchainQueries()
+    //              .transactionByBlockHashAndIndex(
+    //                  Hash.fromHexString(blockHashOrNumber), transactionIndex);
+    //      blockHash = optional.map(tx -> tx.getBlockHash()).orElse(null);
+    //    }
+    //    return optional
+    //        .map(
+    //            transactionWithMetadata ->
+    //                (context
+    //                    .getBlockReplay()
+    //                    .afterTransactionInBlock(
+    //                        blockHash,
+    //                        transactionWithMetadata.getTransaction().hash(),
+    //                        (transaction, blockHeader, blockchain, worldState,
+    // transactionProcessor) ->
+    //                            extractStorageAt(request, accountAddress, startKey, limit,
+    // worldState))
+    //                    .orElseGet(() -> new JsonRpcSuccessResponse(request.getId(), null))))
+    //        .orElseGet(() -> new JsonRpcSuccessResponse(request.getId(), null));
```

Comment:
Intentionally left commented?

Reply:
intentionally left commented.  I still need to do the intra-block versions.